File: Samples/Beyond.NET.Sample.Managed/Source/NullabilityTests.cs
namespace Beyond.NET.Sample;

public class NullabilityTests
{
    public NullabilityTests()
    {
        
    }
    
    public NullabilityTests(bool throwAnException)
    {
        if (throwAnException) {
            throw new Exception("Expected Exception");
        }
    }

    public static NullabilityTests CreateInstance(bool throwAnException)
    {
        return new NullabilityTests(throwAnException);
    }
    
    public string NonNullableStringProperty { get; set; } = "Hello";
    public string NonNullableStringField = "Hello";

    public string? NullableStringProperty { get; set; } = null;
    public string? NullableStringField = null;

    public string MethodWithNonNullableStringReturnValue() => "Hello";
    public string? MethodWithNullableStringReturnValue() => null;

    public string MethodWithNonNullableStringParameter(string value)
    {
        if (value == null) {
            throw new ArgumentNullException(nameof(value));
        }

        return value;
    }
    
    public string MethodWithNonNullableStringParameterThatThrows(string value)
    {
        throw new Exception("Expected Exception");
    }
    
    public string? MethodWithNullableStringParameter(string? value)
    {
        return value;
    }
}

File: Samples/Beyond.NET.Sample.Swift/Tests/TestClasses/NullabilityTests.swift
import XCTest
import BeyondDotNETSampleKit

final class NullabilityTests: XCTestCase {
    override class func setUp() {
        Self.sharedSetUp()
    }
    
    override class func tearDown() {
        Self.sharedTearDown()
    }
    
    func testMethodPropertyAndFieldNullability() throws {
        let test = try XCTUnwrap(Beyond.NET.Sample.NullabilityTests())
        
        let helloString = "Hello".dotNETString()
        let nilString: System.String? = nil
        
        XCTAssertEqual(try test.methodWithNonNullableStringParameter(helloString), helloString)
        XCTAssertEqual(try test.methodWithNullableStringParameter(nilString), nilString)
        
        XCTAssertEqual(try test.nonNullableStringProperty, helloString)
        XCTAssertEqual(test.nonNullableStringField, helloString)
        XCTAssertEqual(try test.nullableStringProperty, nilString)
        XCTAssertEqual(test.nullableStringField, nilString)
        
        XCTAssertEqual(try test.methodWithNonNullableStringReturnValue(), helloString)
        XCTAssertEqual(try test.methodWithNullableStringReturnValue(), nilString)
    }
    
    func testConstructor() throws {
        let _ = try Beyond.NET.Sample.NullabilityTests(false) // Should not throw
        XCTAssertThrowsError(try Beyond.NET.Sample.NullabilityTests(true)) // Should throw
    }
}
